micrometer-core/src/test/java/io/micrometer/core/instrument/push/PushMeterRegistryTest.java

```diff
@@ -79,17 +77,99 @@ void whenUncaughtExceptionInPublish_taskStillScheduled() throws InterruptedExcep
 
     @Test
     void whenUncaughtExceptionInPublish_closeRegistrySuccessful() {
+        PushMeterRegistry pushMeterRegistry = new ThrowingPushMeterRegistry(config, latch);
         assertThatCode(() -> pushMeterRegistry.close()).doesNotThrowAnyException();
     }
 
     @Test
     @Issue("#3712")
     void publishOnlyHappensOnceWithMultipleClose() {
-        pushMeterRegistry = new CountingPushMeterRegistry(config, Clock.SYSTEM);
+        CountingPushMeterRegistry pushMeterRegistry = new CountingPushMeterRegistry(config, Clock.SYSTEM);
         pushMeterRegistry.close();
-        assertThat(((CountingPushMeterRegistry) pushMeterRegistry).publishCount.get()).isOne();
+        assertThat(pushMeterRegistry.publishCount.get()).isOne();
         pushMeterRegistry.close();
-        assertThat(((CountingPushMeterRegistry) pushMeterRegistry).publishCount.get()).isOne();
+        assertThat(pushMeterRegistry.publishCount.get()).isOne();
+    }
+
+    @Test
+    @Issue("#3711")
+    void scheduledPublishOverlapWithPublishOnClose() throws InterruptedException {
+        MockClock clock = new MockClock();
+        CyclicBarrier barrier = new CyclicBarrier(2);
```

Comment:
Since the barrier is not reset, is it used because of its "all-or-nothing breakage model" and because it throws a timeout?
I guess that's also why the result of `await` can be ignored.

Since no reset is needed I believe I would used a `CountDownLatch` and checking the result of `await` to see it this was normal termination or an error, but I think using a barrier makes this simpler. 👍🏼

Reply:
I originally tried using a CountDownLatch, but it doesn't ensure we're checking what we want to check. The same thread could count down twice, so I thought it safer to use the barrier to ensure we had two different threads waiting at the same time. I'm happy to consider ways to make the test better. Feel free to follow-up post merge. It's possible the test won't catch all kinds of possible failures, but I did check that it failed before the changes to main code at least.